CodeCamp #15

NoneLand opened the pull request:
## Motivation

This PR adds sigmoid focal loss CPU implementation on the basis of CUDA implementation.

## Modification

* Add sigmoid focal loss cpu implementation

* Add focal loss unit test on cpu

* Add ops to EN/ZH documents

## Checklist

**Before PR**:

- [x] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) to create this PR.
- [x] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) are used to fix the potential lint issues.
- [ ] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [x] New functionalities are covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [x] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with some of those projects, like MMDet or MMCls.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

grimoire commented on mmcv/ops/csrc/pytorch/cpu/sigmoid_focal_loss.cpp:
```diff
@@ -0,0 +1,104 @@
+                                      Tensor weight, Tensor grad_input,
+                                      float gamma, float alpha);
+
+REGISTER_DEVICE_IMPL(sigmoid_focal_loss_forward_impl, CPU, sigmoid_focal_loss_forward_cpu);
```
As the input type are the same between `TensorSigmoidFocalLossForwardCPUKernelLaucher` and `sigmoid_focal_loss_forward_cpu`, we can remote `sigmoid_focal_loss_forward_cpu` and register `TensorSigmoidFocalLossForwardCPUKernelLaucher`.
Same for `TensorSigmoidFocalLossBackwardCPUKernelLaucher`

NoneLand replied:
Thanks for your review opinion. I would fix this because it saves code.

grimoire commented on mmcv/ops/csrc/pytorch/cpu/sigmoid_focal_loss.cpp:
```diff
@@ -0,0 +1,104 @@
+                               const float alpha,
+                               const int num_classes) {
+    for (int i = 0; i < N; i++) {
+        T p = (T) 1. / ((T)1. + expf(-input[i]));
```
expf can be replaced with exp since c++ has overloaded it.

NoneLand replied:
OK. I would fix this in next commit.

grimoire reviewed (commented):
Since this op can be implemented with native torch. Please provide a benchmark between this ops and torch implementation.

NoneLand commented:
Thanks for reviewer's opinion. I have some question about the benchmark. Could you please help me figure out?

1. Is there a benchmark example for this op? I think this answer would be YES due to the existing cuda op.
2. I would like to know the goal for mmcv to reimplement this op? I would guess the goal is performance due to the benchmark requirement. If the benchmark result shows current op impl has no superior performance than torch impl, would merge delay until the optimization is completed or would merge before optimization?(HAHA, codecamp has a ddl^_^)
3. If there was no benchmark before, here is my plan, could you please help me check? I would test for different batch (2, 4, 8, ..., 8192) and different num_classes (2, 4, 8, ..., 8192) for the python api, each case run 10 times, and cal the average latency of both impl.

Thanks again^_^

grimoire commented:
1. Yes, it would help us to find the bottleneck of the ops.
2. It would be cool if the benchmark can guide you to improve the performance of the op. And we will merge this PR even when the performance is not the best (less is better than nothing). Your benchmark would help us to optimize it in the future.
3. The plan sounds cool. It would be even better if you can perform some tests on the data size of the real scene. [pytest-benchmark](https://pypi.org/project/pytest-benchmark/) or [PyTorch profiler](https://pytorch.org/tutorials/recipes/recipes/profiler_recipe.html) can help you build the benchmark. You can place the benchmark in this PR.

NoneLand commented:
I have used pytest-benchmark to benchmark the implemented sigmoid focal loss op.  Benchmark code is in the latest commit. And on my 8 core notebook, conclusions could be drawn as below: 1. In small data size, implemented op is superior. 2. In medium and large data size, implemented op is superior.  The processed benchmark data could be seen in [benchmark_round10.csv](https://github.com/open-mmlab/mmcv/files/10426594/benchmark_round10.csv) and the raw benchmark data is in 
[0003_4e4fa93e7c5ee5a15dd63068c0fe9992068bc881_20230116_012432_uncommited-changes.txt](https://github.com/open-mmlab/mmcv/files/10426612/0003_4e4fa93e7c5ee5a15dd63068c0fe9992068bc881_20230116_012432_uncommited-changes.txt)

grimoire commented:
Cool.
Do remember to remove the benchmark test code before the final commit. Benchmark test might slow down github action.
